nomos-services/chain/chain-leader/src/leadership.rs

```diff
@@ -124,4 +101,139 @@ impl Leader {
+    sender: &'service Sender<(LeaderPrivate, SecretKey, Epoch)>,
+    last_processed_epoch: Option<Epoch>,
+}
+
+impl<'service> WinningPoLSlotNotifier<'service> {
+    pub(super) const fn new(
+        leader: &'service Leader,
+        sender: &'service Sender<(LeaderPrivate, SecretKey, Epoch)>,
+    ) -> Self {
+        Self {
+            leader,
+            sender,
+            last_processed_epoch: None,
+        }
+    }
+
+    pub(super) fn process_epoch(&mut self, epoch_state: &EpochState) {
+        if let Some(last_processed_epoch) = self.last_processed_epoch {
+            if last_processed_epoch == epoch_state.epoch {
+                tracing::trace!("Skipping already processed epoch.");
+                return;
+            } else if last_processed_epoch > epoch_state.epoch {
+                tracing::error!(
+                    "Received an epoch smaller than the last process one. This is invalid."
+                );
+                return;
+            }
+        }
+        tracing::debug!("Processing new epoch: {:?}", epoch_state.epoch);
+
+        self.check_epoch_winning_utxos(epoch_state);
+    }
+
+    fn check_epoch_winning_utxos(&mut self, epoch_state: &EpochState) {
+        let slots_per_epoch = self.leader.config.epoch_length();
+        let epoch_starting_slot: u64 = self
+            .leader
+            .config
+            .epoch_config
+            .starting_slot(&epoch_state.epoch)
+            .into();
+        // Not used to check if a slot wins the lottery.
+        let latest_tree = UtxoTree::new();
+
+        for utxo in &self.leader.utxos {
+            let note_id = utxo.note.id();
+
+            for offset in 0..slots_per_epoch {
+                let slot = epoch_starting_slot
+                    .checked_add(offset)
+                    .expect("Slot calculation overflow.");
+                let secret_key = self.leader.slot_secret_key(slot.into());
+
+                let Some(public_inputs) =
+                    public_inputs_for_utxo(utxo, epoch_state, slot.into(), &latest_tree)
+                else {
+                    continue;
+                };
+                if !public_inputs.check_winning(utxo.note.value, note_id, *secret_key.as_fr()) {
+                    continue;
+                }
+                tracing::debug!("Found winning utxo with ID {:?} for slot {slot}", utxo.id());
+
+                let leader_private = self.leader.private_inputs_for_winning_utxo_and_slot(
+                    utxo,
+                    epoch_state.epoch,
+                    public_inputs,
+                );
+
+                if let Err(err) =
+                    self.sender
+                        .send((leader_private, secret_key.clone(), epoch_state.epoch))
+                {
+                    tracing::error!(
+                        "Failed to send pre-calculated PoL winning slots to receivers. Error: {err:?}"
+                    );
+                }
+            }
```

Comment:
update the`self.last_processed_epoch` and early return here after we find the first winning slot?

Reply:
In the chain leader subscribe API, I was thinking to return ALL winning slots for a new epoch, instead of making it Blend specific and only return the first one. Since this is not expensive, I thought it would not matter. What do you think?

Reply:
Yes I think we can keep it more generic so any other service can also consume this info if they need to.